### ISSUE TITLE
Broken cross-file references

### ISSUE THREAD

**Issue by SuperAuguste** [zigtools/zls#1071]
`references.zig` uses `DocumentStore.collectDependencies` which collects the imports a file specifies; this incorrect - we should be collecting dependents (files that import the current file), not dependencies (files that the current file imports).

To take this on you'd have to create a new `collectDependents` function that loads in all files in a workspace while checking for imports and returns all files that import the current file.

**Comment by Jarred-Sumner** [comment]
Yes I think this is what I meant with this issue: https://github.com/zigtools/zls/issues/1070

After a recent change, an import 3-4 levels deep through usingnamespace stops working

https://github.com/oven-sh/bun/blob/6f59bc7d2e7aa64956bdd6f56769467853fd0242/src/main.zig#L45

**Comment by techatrix** [comment]
To make it clear what this issue means it practice, here is a symbol references test that is currently failing:

https://github.com/zigtools/zls/blob/fedbf31eb7d7cb55888e656501767cbe423061af/tests/lsp_features/references.zig#L208-L216

The first file declares a public struct and the second file tries to import and use it. Asking for symbol references on the struct declaration in the first file will not find the reference in the second file.

**Comment by Argmaster** [comment]
Hi, I am not sure if this belongs here, I couldn't find issue description that would fit my case exactly, but this one sounds a bit like it, so here is the deal:

In project, ZLS has no idea what `const main = @import("root")` means, gives no suggestions, can not resolve the "root" when I try to go to definition. As a result any symbol referenced through `main` variable become completely opaque.  It makes navigating files grouped in folders really troublesome :C

Found in project called [Cubyz](https://github.com/PixelGuys/Cubyz).

![Image](https://github.com/user-attachments/assets/dda14cd0-4fad-4057-9f3a-4eba5832b19f)

Is that duplicate for this issue?
Are there any workarounds for this?

**Comment by RoadRoller01** [comment]
Wouldn't it be better to first search for the symbol in all files within a workspace, and then process only those files if they are really referencing the symbol.

**Comment by RoadRoller01** [comment]
And I it could also be better if the symbol accorded in a file you immediately check if its really referencing it, Instead of collecting an array list then process it.

**Comment by dismint** [comment]
To clarify this issue, is there any case where a `textDocument/references` works across files? I could not reproduce a scenario where it found references that weren't in the document the request was being made from. Seems like this is more of a missing feature than a bug in that case.

**Comment by techatrix** [comment]
> To clarify this issue, is there any case where a `textDocument/references` works across files?

Here is an example:

```zig
// foo.zig
pub const Foo = struct {};
```

```zig
// bar.zig
const foo = @import("foo.zig");
const Foo = foo.Foo;
//              ^^^ request references here
```
